cpp/include/cudf/groupby.hpp

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <utility>
+#include <vector>
+
+namespace cudf {
+namespace experimental {
+
+/**
+ * @brief Interpolation method to use, when the desired quantile lies between
+ * two data points i and j
+ *
+ * TODO: Move this somewhere else once quantiles are ported to libcudf++
+ *
+ */
+struct interpolation {
+  enum type {
+    LINEAR = 0,  ///< Linear interpolation between i and j
+    LOWER,       ///< Lower data point (i)
+    HIGHER,      ///< Higher data point (j)
+    MIDPOINT,    ///< (i + j)/2
+    NEAREST      ///< i or j, whichever is nearest
+  };
+};
+
+namespace groupby {
+/**
+ * @brief Base class for specifying the desired aggregation in an
+ * `aggregation_request`.
+ *
+ * Other kinds of aggregations may derive from this class to encapsulate
+ * additional information needed to compute the aggregation.
+ */
+struct aggregation {
+  /**
+   * @brief Possible aggregation operations
+   */
+  enum Kind { SUM, MIN, MAX, COUNT, MEAN, MEDIAN, QUANTILE };
```

Comment:
We should re-use this enum https://github.com/rapidsai/cudf/pull/3273/files#diff-19d6a8790460cbd8bb1648ee41619aeaR43-R54. Or figure out another place to keep them so that we don't have two different definitions for each aggregation kind.

Reply:
How about this: `Kind` becomes a category. It can have categories corresponding to these: https://docs.google.com/spreadsheets/d/1PyNYv-YPVobIIg44PUPLRKscTfFgIz1Z2BPtynkSgk8/edit#gid=0&range=G:G
Then, derive from aggregation to get `reduction_aggregation` which holds a member `reduction::operator`.

Reply:
We can't reuse the enum because there are enums on that list that are not supported by groupby.

Reply:
I don't know what you mean by "category".

Reply:
> We can't reuse the enum because there are enums on that list that are not supported by groupby.

But all of those are eventually going to be supported for groupby right? There can be another way to indicate that even though it exists in the enum, it's not currently supported for groupby, but will be in the future.

And by category, I mean:
```c++
enum Kind { REDUCTION, SCAN, COMPACTION ... };
```